WDL/runtime/workflow.py

```diff
@@ -589,41 +594,59 @@ def run_local_workflow(
     write_values_json(posix_inputs, os.path.join(run_dir, "inputs.json"), namespace=workflow.name)
 
     state = StateMachine(run_id, run_dir, workflow, posix_inputs)
-
-    try:
-        while state.outputs is None:
-            if _test_pickle:
-                state = pickle.loads(pickle.dumps(state))
-
-            next_call = state.step()
-            if next_call:
-                if isinstance(next_call.callee, Tree.Task):
-                    run_callee = run_local_task
-                elif isinstance(next_call.callee, Tree.Workflow):
-                    run_callee = run_local_workflow
-                else:
-                    assert False
-                _, outputs = run_callee(
-                    next_call.callee,  # pyre-fixme
-                    next_call.inputs,
-                    run_id=next_call.id,
-                    run_dir=os.path.join(run_dir, next_call.id),
-                )
-                state.call_finished(next_call.id, outputs)
-    except Exception as exn:
-        logger.debug(traceback.format_exc())
-        if isinstance(exn, TaskFailure):
-            logger.error("%s failed", getattr(exn, "run_id"))
-        else:
-            msg = ""
-            if hasattr(exn, "job_id"):
-                msg += getattr(exn, "job_id") + " "
-            msg += exn.__class__.__name__
-            if str(exn):
-                msg += ", " + str(exn)
-            logger.error(msg)
-            logger.info("run directory: %s", run_dir)
-        raise
+    with TerminationSignalFlag(logger) as terminating:
+        try:
+            with thread_pool as executor:
+                while state.outputs is None:
+                    if _test_pickle:
+                        state = pickle.loads(pickle.dumps(state))
+                    next_call = state.step()
+                    while next_call:
+                        if isinstance(next_call.callee, Tree.Task):
+                            run_callee = run_local_task
+                        elif isinstance(next_call.callee, Tree.Workflow):
+                            run_callee = run_local_workflow
+                        else:
+                            assert False
+                        future = executor.submit(run_callee,
+                                                 next_call.callee,  # pyre-fixme
+                                                 next_call.inputs,
+                                                 run_id=next_call.id,
+                                                 run_dir=os.path.join(run_dir, next_call.id),
+                                                 thread_pool=thread_pool,
+                                                 )
+                        future_task_map[future] = next_call.id
+                        next_call = state.step()
+                    done_iter = futures.as_completed(future_task_map)
+                    future = next(done_iter, None)
+                    if future:
```

Comment:
I think we can add `else: assert state.outputs is not None`, i.e. if we've completed all outstanding calls and `state.step()` yields no more, the workflow should be done. It's just nice to put assertions like this to avoid getting into infinite while loops even if our code is buggy.

Reply:
👍